File: frontend/src/services/api/super-agent-activity.ts
/**
 * v0.7.7: Super-Agent Activity Inspector API client.
 *
 * Wraps three /admin/super-agents endpoints used by the inspector page:
 *   GET /admin/super-agents/{id}/activity   — timeline list (limit/since/types).
 *   GET /admin/super-agents/{id}/rollup     — header-card rollup with status pill.
 *   GET /admin/super-agents/sessions/{session_id}/activity — per-session drill-down.
 */
import { apiFetch } from './client';

export type SuperAgentStatusPill = 'active' | 'errored' | 'idle' | 'healthy';

export interface SuperAgentActivityEvent {
  id: number;
  super_agent_id: string;
  session_id: string | null;
  event_type: string;
  recorded_at: string;
  payload: string;
  cost_tokens_in: number | null;
  cost_tokens_out: number | null;
  cost_usd: number | null;
  status: string;
  error_message: string | null;
  duration_ms: number | null;
}

export interface SuperAgentActivityListResponse {
  events: SuperAgentActivityEvent[];
}

export interface SuperAgentRollup {
  super_agent_id: string;
  event_count: number;
  error_count: number;
  total_cost_usd: number;
  last_active_at: string | null;
  status_pill: SuperAgentStatusPill;
  cost_per_event_avg: number | null;
  error_rate: number | null;
}

export const superAgentActivityApi = {
  list(
    id: string,
    opts: { limit?: number; since?: string; types?: string[] } = {},
  ): Promise<SuperAgentActivityListResponse> {
    const qs = new URLSearchParams();
    if (opts.limit) qs.set('limit', String(opts.limit));
    if (opts.since) qs.set('since', opts.since);
    if (opts.types?.length) qs.set('types', opts.types.join(','));
    const query = qs.toString();
    return apiFetch<SuperAgentActivityListResponse>(
      `/admin/super-agents/${encodeURIComponent(id)}/activity${query ? `?${query}` : ''}`,
    );
  },
  rollup(id: string, windowDays: number = 7): Promise<SuperAgentRollup> {
    return apiFetch<SuperAgentRollup>(
      `/admin/super-agents/${encodeURIComponent(id)}/rollup?window_days=${windowDays}`,
    );
  },
  listForSession(
    sessionId: string,
    limit: number = 200,
  ): Promise<SuperAgentActivityListResponse> {
    return apiFetch<SuperAgentActivityListResponse>(
      `/admin/super-agents/sessions/${encodeURIComponent(sessionId)}/activity?limit=${limit}`,
    );
  },
};

File: backend/app_litestar/routes/super_agent_activity.py
"""v0.7.7: admin endpoints for super-agent activity inspector.

Three GET endpoints power the inspector page:
- ``/admin/super-agents/{id}/activity`` — timeline list with limit/since/types filters.
- ``/admin/super-agents/{id}/rollup`` — header-card stats (event count,
  cost, status pill, error rate).
- ``/admin/super-agents/sessions/{session_id}/activity`` — per-session
  timeline drill-down.

All require admin via ``requires_role("admin")``.
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any

from litestar import Router, get
from litestar.exceptions import HTTPException

from app.services import super_agent_activity_service
from app_litestar.auth_guards import requires_role


@get(
    "/{super_agent_id:str}/activity",
    sync_to_thread=True,
    guards=[requires_role("admin")],
)
def list_activity(
    super_agent_id: str,
    limit: int = 200,
    since: str | None = None,
    types: str | None = None,
) -> dict[str, Any]:
    type_list = [t.strip() for t in types.split(",")] if types else None
    return {
        "events": super_agent_activity_service.list_for_super_agent(
            super_agent_id,
            limit=limit,
            since=since,
            types=type_list,
        )
    }


@get(
    "/{super_agent_id:str}/rollup",
    sync_to_thread=True,
    guards=[requires_role("admin")],
)
def get_rollup(super_agent_id: str, window_days: int = 7) -> dict[str, Any]:
    try:
        r = super_agent_activity_service.rollup(super_agent_id, window_days=window_days)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return asdict(r)


@get(
    "/sessions/{session_id:str}/activity",
    sync_to_thread=True,
    guards=[requires_role("admin")],
)
def list_session_activity(session_id: str, limit: int = 200) -> dict[str, Any]:
    return {"events": super_agent_activity_service.list_for_session(session_id, limit=limit)}


super_agent_activity_router = Router(
    path="/admin/super-agents",
    route_handlers=[list_activity, get_rollup, list_session_activity],
)
